hail/src/main/scala/is/hail/Uploader.scala

```diff
@@ -0,0 +1,139 @@
+
+    // for stack trace
+    val e = new Traceback
+
+    val w = new StringWriter
+    // closes w
+    val stackTrace = using(new PrintWriter(w)) { p =>
+      e.printStackTrace(p)
+      w.toString
+    }
+
+    val contents =
+      s"ir0:\n${ Pretty(ir0) }\n\nir:\n${ Pretty(ir) }\n\nfrom:\n${ stackTrace }"
+
+    theUploader.enqueueUpload("ir", contents, pipelineEmail)
+  }
+
+  def upload(typ: String, contents: String, email: String) {
+    theUploader.upload(typ, contents, email)
+  }
+}
+
+class Uploader { self =>
+
+  private val config = {
+    val hc = HailContext.get
+    val sc = hc.sc
+    val hConf = hc.hadoopConf
+    val runtime = Runtime.getRuntime
+
+    JObject(
+      "hostname" -> JString(InetAddress.getLocalHost.getHostName),
+      "jvm_version" -> JString(System.getProperty("java.version")),
+      "jvm_properties" -> JObject(
+        System.getProperties.asScala.map { case (k, v) =>
+          k -> JString(v)
+        }
+          .toList),
+      "jvm_runtime" -> JObject(
+        "available_processors" -> JInt(runtime.availableProcessors()),
+        "free_memory" -> JInt(runtime.freeMemory()),
+        "total_memory" -> JInt(runtime.totalMemory()),
+        "max_memory" -> JInt(runtime.maxMemory())),
+      "spark_version" -> JString(hc.sc.version),
+      "spark_conf" -> JObject(
+        sc.getConf.getAll.map { case (k, v) =>
+          k -> JString(v)
+        }
+          .toList),
+      "hadoop_conf" -> JObject(
+        hConf.iterator().asScala.map { entry =>
+          entry.getKey -> JString(entry.getValue)
+        }
+          .toList),
+      "hail_version" -> JString(hc.version))
+  }
+
+  private val httpClient = HttpClientBuilder.create()
+    .setConnectionManager(new PoolingHttpClientConnectionManager())
+    .build()
+
+  private val queue = new LinkedBlockingQueue[(String, String, String)]()
+
+  val t = new Thread {
+    override def run() {
+      while (true) {
+        val item = queue.take()
+        if (item == null)
+          return
+
+        val (typ, contents, email) = item
+
+        try {
+          self.upload(typ, contents, email)
+        } catch {
+          case e: Exception =>
+            warn(s"upload failed, caught $e")
+        }
+      }
+    }
+  }
+
+  t.start()
+
+  def upload(typ: String, contents: String, email: String) {
+    val request = new HttpPost("https://upload.hail.is/upload")
+
+    val jv = JObject(
+      "type" -> JString(typ),
+      "email" -> JString(email),
+      "config" -> config,
+      "contents" -> JString(contents))
+
+    val data = new StringEntity(
+      JsonMethods.compact(JsonMethods.render(jv)))
+    request.addHeader("content-type", "application/json")
+    request.setEntity(data)
+
+    using(httpClient.execute(request)) { response =>
+      EntityUtils.consume(response.getEntity)
+    }
+  }
+
+  def enqueueUpload(typ: String, contents: String, email: String) {
+    queue.offer((typ, contents, email))
+  }
+
+  def join() {
```

Comment:
this isn't used yet, right?

Reply:
Correct.  I debated calling it in `stop()`, but I prefer to think of the enqueued logging as having no cost and not completely reliable.